### PR TITLE
Apply latest version of event by default

### LINKED ISSUE #13161: Append latest version of records by default
With #12764 we can now append different versions of records. By default, we append version 1 of a record. However, that's not easy to maintain. As we add more places where records are appended we have to remember all these and update them when we want to create a new version.

Instead, it makes more sense to append the latest version of a record by default, and allow changing this in specific cases. Adding a new version of a record is then achieved automatically in all places.

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

With this commit we change the behavior of event applying to always apply an event to the latest event applier. This prevents us from always having to specify a version if it's > 1.

Before it would always take version 1 as the default. This works fine and is the latest version for almost all event appliers. However, in the future we will get more and more versioned event appliers. When coding it is nice to not have to worry about versioning. New stuff can always get the latest version. Only when working with old appliers should we have to worry about this and explicitly specify which version  should be used.

In order to achieve this we need to be able to get the latest version in the ResultBuilderBackedEventApplyingStateWriter. In this class we append the record and apply the event applier. As the versions of event appliers is only available in the EventAppliers class I had to add a getLatestVersion method to the interface of this class. Without this we don't know what recordVersion to set on the metadata of the appended record. It is essential that this version matches the actual applied version. If it doesn't we could break our state, as during replay a different applier might be used than during processing.

## Related issues

closes #13161 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by korthout — APPROVED** [review]
Nice! 👏 

👍 This will help add new versions of events more easily

💭 I originally expected to see a registry of record versions so that this could be used for all record types, not just events. But since it's so specifically built into the `TypedEventWriter` it's perfectly reasonable to keep it separate. If we ever need to write new versions of commands/rejections, then we can add something specific to those writers. For commands this could then be based on the registered command processors instead of the registered event appliers 🤔

LGTM 👍 

🔧 Please consider my suggestions before merging (otherwise the `CREATED` event doesn't benefit if we ever introduce version 3 of it)

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/ResultBuilderBackedEventApplyingStateWriter.java [hunk 1]
```diff
@@ -38,7 +38,8 @@ public ResultBuilderBackedEventApplyingStateWriter(
 
   @Override
   public void appendFollowUpEvent(final long key, final Intent intent, final RecordValue value) {
-    appendFollowUpEvent(key, intent, value, RecordMetadata.DEFAULT_RECORD_VERSION);
+    final int latestVersion = eventApplier.getLatestVersion(intent);
+    appendFollowUpEvent(key, intent, value, latestVersion);
```
[the remark this answers is not in the document]
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/writers/ResultBuilderBackedEventApplyingStateWriter.java:42]: > Now that the appendFollowUpEvent has different behavior (uses the latest by default), let's use this when writing DeploymentIntent.CREATED in the DeploymentCreateProcessor.

Good catch 👍

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Pull request merged**